### PR TITLE
Processing Data in Models

### PR DESCRIPTION
**Pull request by aeh1707**
In this branch I have done the following:

- [x] Set up associations between models.
- [x] Write SQL queries with Active Record.

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -0,0 +1,9 @@
+class User < ApplicationRecord
+  has_many :posts, foreign_key: 'author_id'
+  has_many :likes, foreign_key: 'author_id'
+  has_many :comments, foreign_key: 'author_id'
+
+  def recent_posts
+    posts.order(created_at: :desc).limit(5)
+  end
```
Comment (fobadara) [inline app/models/user.rb:8]: [REQUIRED] Kindly reduce the recent posts to 3 instead of 5 as required in your project

File: app/models/post.rb [hunk 2]
```diff
@@ -0,0 +1,13 @@
+class Post < ApplicationRecord
+  belongs_to :author, class_name: 'User', foreign_key: 'author_id'
+  has_many :comments
+  has_many :likes
+
+  def user_posts_counter=(count)
+    author.update_attribute 'posts_counter', count
+  end
```
Comment (fobadara) [inline app/models/post.rb:8]: [OPTIONAL] You could have just done this:
```
def update_post_counter
  author.increment!(:posts_counter)
end

```

[OPTIONAL] You can also include an active callback that runs after a new post has been created like this:
`after_save :update_post_counter`

File: app/models/comment.rb [hunk 3]
```diff
@@ -0,0 +1,8 @@
+class Comment < ApplicationRecord
+  belongs_to :author, class_name: 'User', foreign_key: 'author_id'
+  belongs_to :post
+
+  def post_comments_counter=(count)
+    post.update_attribute 'comments_counter', count
+  end
```
Comment (fobadara) [inline app/models/comment.rb:7]: [OPTIONAL] As mentioned for the posts counter you can also us the after_save. method to increment when in a new comment is created

Comment (fobadara) [inline app/models/comment.rb:7]: [OPTIONAL] As mentioned for the posts counter you can also us the after_save  method to increment when in a new comment is created

File: app/models/like.rb [hunk 4]
```diff
@@ -0,0 +1,8 @@
+class Like < ApplicationRecord
+  belongs_to :author, class_name: 'User', foreign_key: 'author_id'
+  belongs_to :post
+
+  def post_likes_counter=(count)
+    post.update_attribute 'likes_counter', count
+  end
```
Comment (fobadara) [inline app/models/like.rb:7]: [OPTIONAL] As mentioned for the posts counter you can also use the after_save method to increment when in a new like is created

**Review by fobadara — CHANGES_REQUESTED** [review]
Hi @aeh1707, 

Good job so far!
You still need to work on some issues to go to the next project, but you are almost there!

## Highlights
✅ Linters are passing

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better._



Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by mikethreels — APPROVED** [review]
Hi Abdelhadi, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights
- linters are passing
## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by MeqdamAlqudah — APPROVED** [review]
Hi @aeh1707 , 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

image

## Highlight 
:+1: Readme file
Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**